### PR TITLE
Add support for large-number suffixing

### PR DESCRIPTION
**Pull request by rich-iannone**
This change makes it possible to apply automatic scaling and suffixing (with a predefined or custom symbol set) to larger numbers—thousands, millions, billions, trillions.

A single argument `suffixing` provides the interface for this transformation in `fmt_number()` and `fmt_currency()`. The argument accepts a `TRUE` or `FALSE` (the default), whereby `TRUE` performs the autoscaling/suffixing with a default set of symbols.

The `suffixing` argument could also accept a 4-element character vector understood to represent replacement symbols for the stock `K`, `M`, `B`, and `T` (`NA` values can be used here to exclude those particular ranges from taking part in the transformation).

By way of example, we can use `countrypops` to build a **gt** table and format all of the columns with `fmt_number()` whilst using `suffixing = TRUE`

```r
countrypops %>%
  dplyr::select(
    country_code_3, year, population) %>%
  dplyr::filter(
    country_code_3 %in% c(
      "CHN", "IND", "USA", "PAK", "IDN")
  ) %>%
  dplyr::filter(year > 1975 & year %% 5 == 0) %>%
  tidyr::spread(year, population) %>%
  dplyr::arrange(desc(`2015`)) %>%
  gt(rowname_col = "country_code_3") %>%
  fmt_number(
    columns = TRUE,
    decimals = 2,
    suffixing = TRUE
  )
```

and this yields the following output table:

suffixing_example

The alternative before the change is to use a very large number of `fmt_number()` statements that each use conditional statements to target cells within fixed ranges before applying numerical scaling and pattern-based formatting to each.

Fixes https://github.com/rstudio/gt/issues/134

### CODE REVIEW

**Comment by r-cheologist** [comment]
Please, extend/modify this functionality to NICE scientific notation, e.g.
`1500000` --> "1.5 <&#215; OR &#183; e.g.> 10^6^

**Comment by rich-iannone** [comment]
@r-cheologist the `fmt_scientific()` function has this covered: https://gt.rstudio.com/reference/fmt_scientific.html.

**Comment by douglaszickuhr** [comment]
Hi! Just loved this package and this function. It is so handy! I would say it would be great to have it working with negatives values as well, what do you think?

``` r
library(gt)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

tribble(
  ~"Name",~"Population",
  "x",172975223,
  "y",11204214000,
  "t",-901202001
) %>% 
  gt() %>%
  fmt_number(
    columns = vars(Population),
    suffixing = TRUE
  ) 
```





<table class="gt_table">

<tr>

<th class="gt_col_heading gt_left" rowspan="1" colspan="1">

Name

</th>

<th class="gt_col_heading gt_right" rowspan="1" colspan="1">

Population

</th>

</tr>

<tbody class="gt_table_body">

<tr>

<td class="gt_row gt_left">

x

</td>

<td class="gt_row gt_right">

172.98M

</td>

</tr>

<tr>

<td class="gt_row gt_left gt_striped">

y

</td>

<td class="gt_row gt_right gt_striped">

11.20B

</td>

</tr>

<tr>

<td class="gt_row gt_left">

t

</td>

<td class="gt_row gt_right">

\-901,202,001.00

</td>

</tr>

</tbody>

</table>





Created on 2019-01-20 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by rich-iannone** [comment]
@douglaszickuhr thanks for pointing this out! Completely missed negative values but, yes, they ought to be included. Thanks much for providing some revised code. The solution with `abs()` seems like the way to go!

**Comment by douglaszickuhr** [comment]
> @douglaszickuhr thanks for pointing this out! Completely missed negative values but, yes, they ought to be included. Thanks much for providing some revised code. The solution with `abs()` seems like the way to go!

Thanks @rich-iannone! Is it OK if I send a PR for fixing that? I really would like to contribute. :)

**Comment by rich-iannone** [comment]
Hi @douglaszickuhr, I could give you credit for this change in this PR by adding you as a co-author to the upcoming commit. Let me know if that's okay with you.

If that is okay, a thing I'd need is for you to sign the [individual](https://rstudioblog.files.wordpress.com/2017/05/rstudio_individual_contributor_agreement.pdf) or [corporate](https://rstudioblog.files.wordpress.com/2017/05/rstudio_corporate_contributor_agreement.pdf) contributor agreement as appropriate. Then, send the signed copy over to jj@rstudio.com.

**Comment by douglaszickuhr** [comment]
@rich-iannone Sure, no problem! I will sign the doc and send that over. What about the PR, should I submit one?

**Comment by rich-iannone** [comment]
@douglaszickuhr Great, thanks for getting that contributor agreement in. No need to file a separate PR, what I'll do momentarily is add you as a co-author in a commit for this PR. This will put you in the contributors list!

File: R/format_data.R [hunk 1]
```diff
@@ -135,6 +153,52 @@ fmt_number <- function(data,
     sep_mark <- ""
   }
 
+  # Determine whether large-number suffixing is being
+  # used and set the appropriate inputs
+  if (is.logical(suffixing) &&
+      length(suffixing) == 1 &&
+      suffixing) {
```
Comment (jcheng5) [inline R/format_data.R]: `isTRUE(suffixing)`

File: R/format_data.R [hunk 2]
```diff
@@ -135,6 +153,52 @@ fmt_number <- function(data,
...
+  } else if (is.logical(suffixing) &&
+             length(suffixing) == 1 &&
+             suffixing == FALSE) {
+
+    # If `suffixing` is FALSE, the `to_suffix`
+    # logical will also be set to FALSE
+    to_suffix <- FALSE
+
+  } else if (is.character(suffixing) &&
+             length(suffixing) == 4){
```
Comment (jcheng5) [inline R/format_data.R]: Check that vector is unnamed (so we can use names later to mean other things)

Comment (rich-iannone) [reply R/format_data.R]: I used the following to `stop()` even if there are partially-named vectors:

```r
} else if (is.character(suffixing) &&
             length(suffixing) == 4){

    # Stop function if the character vector `suffixing`
    # contains any names
    if (!is.null(names(suffixing))) {
      stop("The character vector supplied to `suffixed` cannot contain names.",
           call. = FALSE)
    }

...
}
```

File: R/format_data.R [hunk 3]
```diff
@@ -135,6 +153,52 @@ fmt_number <- function(data,
...
+    stop("The value provided to `suffixing` must either be:\n",
+         " * `TRUE` or `FALSE` (the default)\n",
+         " * a four-element character vector with suffixing text",
+         call. = FALSE)
+  }
+
+  # If choosing to perform large-number suffixing
+  # of numeric values, force `scale_by` to be 1.0
+  if (to_suffix) {
+    scale_by <- 1.0
```
[the remark this answers is not in the document]
Comment (rich-iannone) [reply R/format_data.R]: I added this possibility for a warning before forcing `scale_by` to `1.0`:

```r
  if (to_suffix) {

    if (!missing(scale_by) & !identical(scale_by, 1.0)) {
      warning(...)
    }

    scale_by <- 1.0
  }
```

**Comment by rich-iannone** [comment]
> Is there an edge case with 999999.999999 and `decimals = 2`? (Seems like you'd want 1M, but you'd get 1000K?)

This has now been addressed in the most recent commits. Tests were added as well.

**Comment by rich-iannone** [comment]
> Please refactor to eliminate repeated code

Repeated statements were extracted as functions and now those large passages have been replaced with function calls.

File: R/format_data.R [hunk 4]
```diff
@@ -135,6 +153,23 @@ fmt_number <- function(data,
...
   }
 
+  # Get the `suffixing_inputs`
+  suffixing_inputs <- get_suffixing_inputs(suffixing)
+
+  # If choosing to perform large-number suffixing
+  # of numeric values, force `scale_by` to be 1.0
+  if (suffixing_inputs$to_suffix) {
+
+    if (!missing(scale_by) & !identical(scale_by, 1.0)) {
```
Comment (jcheng5) [inline R/format_data.R]: The `&` should be `&&`. The single-`&` is only when you want to vectorize over whole vectors of logicals, not when you're trying to AND two single-element logical vectors.

Comment (rich-iannone) [reply R/format_data.R]: This has been corrected.

File: R/utils.R [hunk 5]
```diff
@@ -384,6 +384,118 @@ get_pre_post_txt <- function(pattern) {
...
+
+  } else if (is_x_false(suffixing)) {
+
+    # If `suffixing` is FALSE, the `to_suffix`
+    # logical will also be set to FALSE
+    to_suffix <- FALSE
+    num_suffixes <- NA
+
+  } else if (is.character(suffixing) &&
+             length(suffixing) == 4){
```
[the remark this answers is not in the document]
Comment (rich-iannone) [reply R/utils.R]: This has been entirely removed.

File: R/utils.R [hunk 6]
```diff
@@ -384,6 +384,118 @@ get_pre_post_txt <- function(pattern) {
...
+    isFALSE(x)
+
+  } else {
+
+    is.logical(x) && length(x) == 1L && !is.na(x) && !x
+  }
+}
+
+# Get large-number suffixing inputs
+get_suffixing_inputs <- function(suffixing) {
```
[the remark this answers is not in the document]
Comment (rich-iannone) [reply R/utils.R]: This has been fixed.

**Comment by rich-iannone** [comment]
Everything is now ready for a final look. Additional tests were written to catch some edge cases around `NA` placement in a character vector supplied to `suffixing`. Documentation for the `suffixing` argument was edited to correspond to the new behavior.

File: R/format_data.R [hunk 7]
```diff
@@ -135,6 +156,23 @@ fmt_number <- function(data,
     sep_mark <- ""
   }
 
+  # Get the `suffixing_inputs`
+  suffixing_inputs <- get_suffixing_inputs(suffixing)
```
Comment (rich-iannone) [inline R/format_data.R]: Rename function as `normalize_***()` and return as a vector.

File: R/utils.R [hunk 8]
```diff
@@ -384,6 +384,166 @@ get_pre_post_txt <- function(pattern) {
...
+
+    isFALSE(x)
+
+  } else {
+
+    is.logical(x) && length(x) == 1L && !is.na(x) && !x
+  }
+}
+
+# Get large-number suffixing inputs
```
Comment (rich-iannone) [inline R/utils.R]: Improve comment

File: R/utils.R [hunk 9]
```diff
@@ -384,6 +384,166 @@ get_pre_post_txt <- function(pattern) {
...
+
+  # Determine whether large-number suffixing is being
+  # used and set the appropriate inputs
+  if (isTRUE(suffixing)) {
+
+    # If `suffixing` is TRUE, the `num_suffixes`
+    # logical will also be set to TRUE and the
+    # default set of symbols will be assigned
+    # to `num_suffixes`
+    num_suffixes <- c("K", "M", "B", "T")
```
Comment (rich-iannone) [inline R/utils.R]: Return early at each block

File: R/utils.R (hunk 8 above)
Comment (rich-iannone) [reply R/utils.R]: This is now much improved (in https://github.com/rstudio/gt/pull/163).

File: R/utils.R (hunk 9 above)
Comment (rich-iannone) [reply R/utils.R]: This function has been rewritten (https://github.com/rstudio/gt/pull/163).

File: R/format_data.R (hunk 7 above)
Comment (rich-iannone) [reply R/format_data.R]: Changed in https://github.com/rstudio/gt/pull/163.
